File: algorithm/课件/第14课 课件 代码/lc212 单词搜索 II.cpp
class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        root = new Node();
        for (string& word : words) insert(word);
        for (int i = 0; i < board.size(); i++)
            for (int j = 0; j < board[0].size(); j++) {
                used = vector<vector<bool>>(board.size(), vector<bool>(board[0].size(), false));
                used[i][j] = true;
                dfs(board, i, j, root);
            }
        return ans;
    }

private:
    struct Node {
        string* word;
        unordered_map<char, Node*> child;
    };
    Node* root;
    vector<vector<bool>> used;
    vector<string> ans;
    const int dx[4] = {-1, 0, 0, 1};
    const int dy[4] = {0, -1, 1, 0};

    void dfs(vector<vector<char>>& board, int x, int y, Node* curr) {
        char ch = board[x][y];
        if (curr->child.find(ch) == curr->child.end()) return;
        Node* fa = curr;
        curr = fa->child[ch];
        if (curr->word != nullptr) {
            ans.push_back(*curr->word);
            curr->word = nullptr;
        }
        if (curr->child.empty()) {
            fa->child.erase(ch);
        }
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || ny < 0 || nx >= board.size() || ny >= board[0].size()) continue;
            if (used[nx][ny]) continue;
            used[nx][ny] = true;
            dfs(board, nx, ny, curr);
            used[nx][ny] = false;
        }
    }

    void insert(string& word) {
        Node* curr = root;
        for (char c : word) {
            if (curr->child[c] == nullptr) {
                curr->child[c] = new Node();
            }
            curr = curr->child[c];
        }
        curr->word = &word;
    }
};

File: algorithm/课件/第14课 课件 代码/lc212 单词搜索 II.java
import java.util.HashMap;

class Solution {
    public class Node {
        String word; // 实际单词的引用
        HashMap<Character, Node> child; // 出边（字符映射）
        Node() {
            word = null;
            child = new HashMap<Character, Node>();
        }
    }
    Node root;

    boolean[][] visited;
    int n;
    int m;
    List<String> ans;

    int[] dx = new int[]{-1, 0, 0, 1};
    int[] dy = new int[]{0, -1, 1, 0};

    public List<String> findWords(char[][] board, String[] words) {
        // 建立字典树，存单词
        root = new Node();
        for (String word : words) {
            insert(word);
        }

        m = board.length;
        n = board[0].length;
        ans = new ArrayList<String>();
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++) {
                visited = new boolean[m][n];
                visited[i][j] = true;
                dfs(board, i, j, root);
            }
        return ans;
    }

    private void dfs(char[][] board, int x, int y, Node curr) {
        char ch = board[x][y];
        if (!curr.child.containsKey(ch)) return;
        Node fa = curr;
        curr = fa.child.get(ch);
        if (curr.word != null) {
            ans.add(curr.word);
            curr.word = null;
        }
        if (curr.child.isEmpty()) {
            fa.child.remove(ch);
        }
        for (int k = 0; k < 4; k++) {
            int nx = x + dx[k];
            int ny = y + dy[k];
            if (nx < 0 || ny < 0 || nx >= m || ny >= n) continue;
            if (visited[nx][ny]) continue;
            visited[nx][ny] = true;
            dfs(board, nx, ny, curr);
            visited[nx][ny] = false;
        }
    }

    private void insert(String word) {
        Node curr = root;
        for (char ch : word.toCharArray()) {
            if (!curr.child.containsKey(ch)) {
                curr.child.put(ch, new Node());
            }
            curr = curr.child.get(ch);
        }
        curr.word = word;
    }
}